src/pallene/pallenelib.lua

```diff
@@ -90,12 +127,209 @@ static lua_Number  pallene_math_log(lua_Integer x, lua_Integer base);
+}
+
+static int pallene_tracer_debug_traceback(lua_State *L) {
+    /* The debug traceback function frame. */
+    pt_frame_t self = {
+        .fn_name  = "pallene_tracer_debug_traceback",
+        .mod_name = "pallene_debug",
+        .line     = 0
+    };
+    pallene_tracer_frameenter(L, &self, (ptrdiff_t) pallene_tracer_debug_traceback);
+
+    const char *message = lua_tostring(L, 1);
+    fprintf(stderr, "Runtime error: %s\nStack traceback: \n", message);
+
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    pt_frame_t *frame = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    /* For context switch from Pallene to Lua or vice versa. */
+    /* We use the respective call-stack depending on the context.
+     * In Lua context we use the Lua call-stack and in Pallene context
+     * we use our self-maintained call stack. */
+    /* 1: Lua, 0: Pallene */
+    int context = 1;
+
+    /* Current level of depth we are at in the Lua call stack. */
+    int level   = 0;
+
+    /* Are we done iterating through all the call-frames in the Lua stack? */
+    bool gstack  = 0;
+
+    /* Which context we were in previously? */
+    /* The numeric representation is same as `context`. */
+    int prev_context = 1;
+
+    /* The frame signature. */
+    ptrdiff_t frame_sig = frame -> frame_sig;
+
+    while(true) {
+        /* If we are done iterating through the Lua stack and
+           there is no more frame left in Pallene Tracer stack,
+           we are done with the traceback. */
+        if(!gstack && frame == NULL)
+            break;
+
+        if(context) {
+            /* Get lua call stack information. */
+            lua_Debug ar;
+
+            if(!(gstack = lua_getstack(L, level, &ar)))
+                continue;
+
+            level++;
+
+            /* We need more info for a good traceback entry. */
+            lua_getinfo(L, "Slnt", &ar);
+
+            /* We have got a C frame. Time to make a context switch. */
+            if(ar.what && !strcmp("C", ar.what)) {
+                lua_getinfo(L, "f", &ar);
+                frame_sig = (ptrdiff_t) lua_tocfunction(L, -1);
+                lua_pop(L, 1);
+
+                context = 0;
+            } else {
+                fprintf(stderr, "    %s:%d: in function '%s'\n", ar.short_src, ar.currentline,
+                        ar.name != NULL ? ar.name : "<anonymous>");
+            }
+
+            prev_context = 1;
+        } else {
+            /* If the frame signature does not match, then it's just a normal
+               C function  */
+            if(frame == NULL || frame -> frame_sig != frame_sig) {
+                /* If we switched from Lua and the frame signature is
+                   not known, then function is just a C function oblivious to Pallene and Lua. */
+                if(prev_context == 1)
+                    fprintf(stderr, "    C Function: 0x%lx\n", (void *) frame_sig);
+
+                context = 1;
+                goto pallene_stack_done;
```

Comment:
I don't like this goto... Looks like it's there because of the prev_context. Perhaps when we get rid of frame_sig, the prev_context will also go away at the same time?

Reply:
We won't/can't get rid of the frame signature for now. Otherwise we won't be able to figure out when to switch to Lua stack especially if there is a C call that is not of Pallenes or is not using our tracer.

But I will try to get rid of the nasty `goto` :)

Reply:
I'm confused about the won't/can't. I have the impression that we can, by giving a pt_frame to the Lua entry points, but maybe you just want to do it on a future PR?

Reply:
We could certainly have a separate type of pt_frame for Lua entry points. But the `frame signature` in our case is important when we are trying to figure out whether we have a C sitting in the middle which Pallene tracer have no idea about. The only way we can figure this out is in my opinion is if we have some sort of signature or essence of the Previous Lua entry point, which is common in both Lua and Pallene stack. Hope that make sense.

Alternatively, you can elaborate your idea even more, that way I will be able to figure out should the frame signature be needed.